lib/googleauth/external_account/aws_credentials.rb

```diff
@@ -0,0 +1,401 @@
+# Copyright 2015 Google, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "time"
+require "googleauth/base_client"
+require "googleauth/helpers/connection"
+require "googleauth/oauth2/sts_client"
+
+module Google
+  # Module Auth provides classes that provide Google-specific authorization
+  # used to access Google APIs.
+  module Auth
+    # Authenticates requests using External Account credentials, such
+    # as those provided by the AWS provider.
+    module ExternalAccount
+      # This module handles the retrieval of credentials from Google Cloud
+      # by utilizing the AWS EC2 metadata service and then exchanging the
+      # credentials for a short-lived Google Cloud access token.
+      class AwsCredentials
+        include BaseClient
+        include Helpers::Connection
+        extend CredentialsLoader
+
+        STS_GRANT_TYPE = "urn:ietf:params:oauth:grant-type:token-exchange".freeze
+        STS_REQUESTED_TOKEN_TYPE = "urn:ietf:params:oauth:token-type:access_token".freeze
+        IAM_SCOPE = ["https://www.googleapis.com/auth/iam"].freeze
+
+        def initialize options = {}
+          @audience = options[:audience]
+          @scope = options[:scope] || IAM_SCOPE
+          @subject_token_type = options[:subject_token_type]
+          @token_url = options[:token_url]
+          @credential_source = options[:credential_source] || {}
+          @service_account_impersonation_url = options[:service_account_impersonation_url]
+          @environment_id = @credential_source["environment_id"]
+          @region_url = @credential_source["region_url"]
```

Comment:
AWS specific URLs should be validated, see https://github.com/googleapis/google-auth-library-java/pull/1079